core/src/test/java/org/apache/iceberg/TestBaseIncrementalAppendScan.java

```diff
@@ -51,6 +51,165 @@ public void testFromSnapshotInclusive() {
+                          snapshotBranchB
+                            \
+                             \
+                              \files:FILE_C
+                          snapshotBranchC(branch:b1)
+    */
+    IncrementalAppendScan scan = newScan().fromSnapshotInclusive(tagSnapshotAName);
+    Assertions.assertThat(scan.planFiles()).hasSize(5);
+
+    IncrementalAppendScan scan2 =
+        newScan().fromSnapshotInclusive(tagSnapshotAName).useBranch(branchName);
+    Assertions.assertThat(scan2.planFiles()).hasSize(3);
+
+    IncrementalAppendScan scan3 = newScan().toSnapshot(snapshotBranchBId).useBranch(branchName);
+    Assertions.assertThat(scan3.planFiles()).hasSize(2);
+
+    IncrementalAppendScan scan4 = newScan().toSnapshot(snapshotBranchCId).useBranch(branchName);
+    Assertions.assertThat(scan4.planFiles()).hasSize(3);
+
+    IncrementalAppendScan scan5 =
+        newScan()
+            .fromSnapshotExclusive(tagSnapshotAName)
+            .toSnapshot(snapshotBranchBId)
+            .useBranch(branchName);
+    Assertions.assertThat(scan5.planFiles()).hasSize(1);
+  }
+
+  @Test
+  public void testUseBranchWithTagShouldFail() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotAId = table.currentSnapshot().snapshotId();
+    String tagSnapshotAName = "t1";
+    table.manageSnapshots().createTag(tagSnapshotAName, snapshotAId).commit();
+
+    Assertions.assertThatThrownBy(
+            () -> newScan().fromSnapshotInclusive(snapshotAId).useBranch(tagSnapshotAName))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage(String.format("Ref %s is not a branch", tagSnapshotAName));
+  }
+
+  @Test
+  public void testUseBranchWithInvalidEndSnapshotShouldFail() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotAId = table.currentSnapshot().snapshotId();
+
+    String branchName = "b1";
+    String tagSnapshotAName = "t1";
+    table.manageSnapshots().createBranch(branchName, snapshotAId).commit();
+    table.manageSnapshots().createTag(tagSnapshotAName, snapshotAId).commit();
+
+    String tagName2 = "t2";
+    table.newFastAppend().appendFile(FILE_B).appendFile(FILE_B).commit();
+    long snapshotMainBId = table.currentSnapshot().snapshotId();
+    table.manageSnapshots().createTag(tagName2, snapshotMainBId).commit();
+
+    table.newFastAppend().appendFile(FILE_C).toBranch(branchName).commit();
+
+    /*
+
+          files:FILE_A            files:FILE_B FILE_B
+     ---- snapshotA(tag:t1) ---- snapshotMainB(tag:t2)
+                        \
+                         \
+                          \files:FILE_C
+                          snapshotBranchB(branch:b1)
+    */
+    Assertions.assertThatThrownBy(
+            () -> newScan().toSnapshot(snapshotMainBId).useBranch(branchName).planFiles())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("End snapshot is not a valid snapshot on the current branch");
+  }
```

Comment:
I think this should also test
```
Assertions.assertThatThrownBy(
            () ->
                newScan().fromSnapshotInclusive(snapshotMainBId).useBranch(branchName).planFiles())
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessageContaining(
            String.format(
                "Starting snapshot (inclusive) %s is not an ancestor of end snapshot 3",
                snapshotMainBId));
```
either in this test method or in a separate one